engines/sword1/screen.cpp

```diff
@@ -1232,7 +1313,41 @@ void Screen::spriteClipAndSet(uint16 *pSprX, uint16 *pSprY, uint16 *pSprWidth, u
 }
 
 void Screen::fnFlash(uint8 color) {
-	warning("stub: Screen::fnFlash(%d)", color);
+	const byte *targetColor = _white;
+
+	switch (color) {
+	case FLASH_RED:
+		targetColor = _red;
+		break;
+	case FLASH_BLUE:
+		targetColor = _blue;
+		break;
+	case BORDER_YELLOW:
+		targetColor = _yellow;
+		break;
+	case BORDER_GREEN:
+		targetColor = _green;
+		break;
+	case BORDER_PURPLE:
+		targetColor = _purple;
+		break;
+	case BORDER_BLACK:
+		targetColor = _black;
+		break;
+	default:
+		warning("Screen::fnFlash(%d): Bogus color", color);
+		return;
+	}
+
+	_system->getPaletteManager()->setPalette(targetColor, 0, 1);
```

Comment:
Do you know where in the game this function is called?

The Mac version of the game is known to use color 255 instead of 0 in some places, for example for the transparent color in parallax, or the background color visible in some screens, such as when opening the door in Nico's apartment.

Reply:
This function was previously unimplemented as it was called (mostly) in debug opcodes (and that comes with the intention of implementing debug stuff in the future). I said "mostly" because there's an actual occurence of `FN_blue_flash()` in what seems to be a define macro for the PSX version of `FN_walk` which I haven't yet checked if we have in our system or if it's actually needed.

I can definitely check a Mac disassembly tonight and let you know.

Reply:
About the Mac issue; I dug into the PPC disassembly and the palette handling seem to be identical to the PC version.

So if I patch `fnSetFadeTargetPalette()` to be like this (sorry, I have to reintroduce the const_cast 🤷‍♂️) it works correctly:

```c++
void Screen::fnSetFadeTargetPalette(uint8 start, uint16 length, uint32 id, bool toBlack) {
	uint8 *rgbData = nullptr;

	if (toBlack) {
		rgbData = const_cast<uint8 *>(_black);
	} else {
		rgbData = (uint8 *)_resMan->openFetchRes(id);
	}

	// Bug #8636: Force color 255 to black
	if (SwordEngine::isMac()) {
		if (start != 0 && start + length == 256)
			rgbData[(length - 1) * 3 + 0] = rgbData[(length - 1) * 3 + 1] = rgbData[(length - 1) * 3 + 2] = 0;
	}

	if (SwordEngine::_systemVars.wantFade) {
		memcpy(_currentPalette + (start * 3), rgbData, length * 3);
		memcpy(_paletteFadeInfo.dstPalette + (start * 3), rgbData, length * 3);
		memset(_paletteFadeInfo.srcPalette + (start * 3), 0, length * 3);
	} else {
		memcpy(_currentPalette + (start * 3), rgbData, length * 3);
	}

	_resMan->resClose(id);
}
```

The solution to this dilemma seems to be to apply your patch here as well, as you can see. If you approve it I will commit that.

Reply:
The test on `start != 0` is I think one of the things that does not make sense in the current code and just works by accident. I can't think of any good reason why I added that check here.

But in any case, I think that rather than add the hack here, it would make more sense to modify `Screen::updateScreen()`:
```diff
void Screen::updateScreen() {
	if (Logic::_scriptVars[NEW_PALETTE]) {
		_updatePalette = true;
		Logic::_scriptVars[NEW_PALETTE] = 0;
	}

	if (_updatePalette) {
		fnSetFadeTargetPalette(0, 184, _roomDefTable[_currentScreen].palettes[0]);
		fnSetFadeTargetPalette(184, 72, _roomDefTable[_currentScreen].palettes[1]);
		fnSetFadeTargetPalette(0, 1, 0, true);
+		// Bug #8636: Force color 255 to black
+		if (SwordEngine::isMac())
+			fnSetFadeTargetPalette(255, 1, 0, true);
		startFadePaletteUp(1);
		_updatePalette = false;
	}
```

Reply:
Thanks, definitely cleaner, love it. Commited!